File: wrappers/csharp/huestream_csharp_integrationtests/helpers/BridgeWrapperImpl.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Net;
using System.IO;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;

namespace huestream_tests
{
    class LightIDImpl : ILightID
    {
        public LightIDImpl(int id, String mac)
        {
            _id = id;
            _mac = mac;
        }

        public int GetLightId()
        {
            return _id;
        }

        public String GetMac()
        {
            return _mac;
        }

        int _id;
        String _mac;
    }

    class BridgeWrapperImpl : IBridgeWrapper
    {
        public BridgeWrapperImpl(String userName, String clientKey, String ip4Address, String bridgeId, String tcpPort)
        {
            _userName = userName;
            _ip4Address = ip4Address;
            _clientKey = clientKey;
            _bridgeId = bridgeId;
            _tcpPort = tcpPort;
        }

        public void RemoveEntertainmentGroups()
        {
            List<int> groups = GetAllEntertainmentGroups();

            if (groups.Count == 0)
                return;

            foreach (int groupId in groups)
            {
                String url = BuildApiUrl("groups/" + groupId.ToString());
                JArray response = Network.PerformDeleteRequest(url);
                GetSuccessNode(response);
            }
        }

        public int GetEntertainmentGroupId()
        {
            List<int> groupIds = GetAllEntertainmentGroups();

            if (groupIds.Count == 0)
            {
                return CreateEntertainmentGroup();
            }
            else
            {
                return groupIds[0];
            }
        }

        public String GetLightRGBColor(ILightID lightId)
        {
            String stipUrl = "http://" + _ip4Address + ":" + _tcpPort + "/stip/devices/" + lightId.GetMac();
            JObject response = Network.PerformGetRequest(stipUrl);

            Assert.IsNotNull(response, "Stip Light status response is null");

            JObject stateNode = (JObject)response["state"];
            Assert.IsNotNull(stateNode, "state node is null");

            String rgbText = (String)stateNode["rgb"];
            Assert.IsNotNull("RGB text value is null", rgbText);

            return rgbText;
        }

        public List<ILightID> GetLLCLightsIDs()
        {
            JObject lightsResponse = Network.PerformGetRequest(BuildApiUrl("lights"));
            var result = new List<ILightID>();

            foreach (var keyValue in lightsResponse)
            {
                String lightId = keyValue.Key;
                var lightInfo = (JObject)keyValue.Value;

                String modelId = (String)lightInfo["modelid"];
                String uniqueId = (String)lightInfo["uniqueid"];
                uniqueId = uniqueId.Substring(0, 23);

                if (modelId.Contains("LLC"))
                {
                    result.Add(new LightIDImpl(int.Parse(lightId), uniqueId));
                }
            }

            return result;
        }

        public void IncludeLightsIntoGroup(List<ILightID> lights, int targetGroupId)
        {
            JArray lightsArray = new JArray();
            foreach (ILightID id in lights)
            {
                lightsArray.Add(id.GetLightId().ToString());
            }

            JObject request = new JObject
            {
                ["lights"] = lightsArray
            };

            JArray response = Network.PerformUpdateRequest(BuildApiUrl("groups/" + targetGroupId), request, Network.UPDATE_REQUEST.PUT);
            GetSuccessNode(response);
        }

        public void SetLightsCoordinates(int groupId, List<ILightCoordinate> lightCoordinates)
        {
            var locationsNode = new JObject();

            foreach (ILightCoordinate coordinate in lightCoordinates)
            {
                JArray coordArray = new JArray();
                coordArray.Add(coordinate.GetX());
                coordArray.Add(coordinate.GetY());
                coordArray.Add(0);

                locationsNode[coordinate.GetLightId().ToString()] = coordArray;
            }

            var request = new JObject
            {
                ["locations"] = locationsNode
            };
            JArray response = Network.PerformUpdateRequest(BuildApiUrl("groups/" + groupId), request, Network.UPDATE_REQUEST.PUT);
            Assert.NotNull(response, "Update locations response is null");

            foreach (var node in response)
            {
                var lightNode = (JObject)node;
                Assert.NotNull(lightNode, "Invalid response format");

                var successNode = (JObject)lightNode["success"];
                Assert.NotNull(successNode, "Creating group was not successful");
            }
        }

        public void CleanUpUser()
        {
            JArray response = Network.PerformDeleteRequest(BuildApiUrl("config/whitelist/" + _userName));
            GetSuccessNode(response);
        }

        int CreateEntertainmentGroup()
        {
            var lightsArray = new JArray{};

            var request = new JObject
            {
                ["type"] = "Entertainment",
                ["lights"] = lightsArray,
                ["class"] = "TV"
            };

            JArray response = Network.PerformUpdateRequest(BuildApiUrl("groups"), request, Network.UPDATE_REQUEST.POST);
            JObject successNode = (JObject)GetSuccessNode(response);

            var groupIDValue = (String)successNode["id"];
            Assert.NotNull(groupIDValue, "Can not find group id");

            return int.Parse(groupIDValue);
        }

        List<int> GetAllEntertainmentGroups()
        {
            JObject groupsResponse = Network.PerformGetRequest(BuildApiUrl("groups"));
            Assert.NotNull(groupsResponse, "Groups JSON object is NULL");

            var result = new List<int>();

            foreach (var keyValue in groupsResponse)
            {
                String groupId = keyValue.Key;
                var group = (JObject)keyValue.Value;

                var groupType = (String)group["type"];
                if (groupType == "Entertainment")
                {
                    result.Add(int.Parse(groupId));
                }
            }

            return result;
        }

        String BuildApiUrl(String urlTail)
        {
            StringBuilder urlBuilder = new StringBuilder();
            urlBuilder.Append("http://")
                      .Append(_ip4Address)
                      .Append(":")
                      .Append(_tcpPort)
                      .Append("/api/")
                      .Append(_userName)
                      .Append("/")
                      .Append(urlTail);

            return urlBuilder.ToString();
        }

        Object GetSuccessNode(JArray response)
        {
            Assert.NotNull(response, "Create/Update/Delete group response is null");

            JToken responseRoot = response[0];
            Assert.NotNull(responseRoot, "Wrong structure of incoming response");

            var successNode = responseRoot["success"];
            Assert.NotNull(successNode, "Creating group was not successful");

            return successNode;
        }

        private readonly String _userName;
        private readonly String _clientKey;
        private readonly String _ip4Address;
        private readonly String _bridgeId;
        private readonly String _tcpPort;
    }
}

File: edk/wrappers/java/android_integration_tests/app/src/androidTest/java/com/lighting/huestream/integrationtests/helpers/BridgeWrapperImpl.java
package com.lighting.huestream.integrationtests.helpers;

import org.json.JSONException;
import org.junit.Assert;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONObject;

class BridgeWrapperImpl implements IBridgeWrapper {
    public BridgeWrapperImpl(final String userName, final String clientKey, final String ip4Address,
                             final String bridgeId, final String tcpPort) {
        _userName = userName;
        _clientKey = clientKey;
        _ip4Address = ip4Address;
        _bridgeId = bridgeId;
        _tcpPort = tcpPort;
    }

    @Override
    public String getLightRBGColor(ILightID lightId) {
        try {
            return getLightRBGColor_internal(lightId);
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }

        return null;
    }

    @Override
    public Integer getEntertainmentGroupId() {
        try {
            return getEntertainmentGroupId_internal();
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }

        return -1;
    }

    @Override
    public void removeEntertainmentGroups() {
        try {
            removeEntertainmentGroups_internal();
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }
    }

    @Override
    public List<ILightID> getLLCLightsIDs() {
        try {
            return getLLCLightsIDs_internal();
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }

        return null;
    }

    @Override
    public void includeLightsIntoGroup(List<ILightID> lights, Integer targetGroupId) {
        try {
            includeLightsIntoGroup_internal(lights, targetGroupId);
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }
    }

    @Override
    public void setLightsCoordinates(Integer groupId, List<ILightCoordinate> lightCoordinates) {
        try {
            setLightsCoordinates_internal(groupId, lightCoordinates);
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }
    }

    @Override
    public void cleanUpUser() {
        try {
            cleanUpUser_internal();
        } catch (JSONException e) {
            Assert.fail("JSONException popped up");
        }
    }

    private void removeEntertainmentGroups_internal() throws JSONException {
        List<Integer> groups = getAllEntertainmentGroups();
        if (!groups.isEmpty()) {

            for(Integer groupId : groups) {
                String url = buildApiUrl("groups/" + groupId.toString());

                JSONArray response = Network.performDeleteRequest(url);
                getSuccessNode(response);
            }
        }
    }

    private String getLightRBGColor_internal(ILightID lightId) throws JSONException {
        final String stipUrl = "http://" + _ip4Address + ":" + _tcpPort + "/stip/devices/" + lightId.getMac();
        final JSONObject response = Network.performGetRequest(stipUrl);
        Assert.assertNotNull("Stip Light status response is null", response);

        final JSONObject stateNode = (JSONObject) response.get("state");
        Assert.assertNotNull("state node is null", stateNode);

        final String rgbText = (String) stateNode.get("rgb");
        Assert.assertNotNull("RGB text value is null", rgbText);

        return rgbText;
    }

    private Integer getEntertainmentGroupId_internal() throws JSONException {
        List<Integer> groupIds = getAllEntertainmentGroups();

        if (groupIds.isEmpty()) {
            return createEntertainmentGroup();
        } else {
            return groupIds.get(0);
        }
    }

    class LightIDImpl implements ILightID {
        public LightIDImpl(Integer id, final String mac) {
            _id = id;
            _mac = mac;
        }

        @Override
        public Integer getLightId() {
            return _id;
        }

        @Override
        public String getMac() {
            return _mac;
        }

        final Integer _id;
        final String _mac;
    }


    private List<ILightID> getLLCLightsIDs_internal() throws JSONException {
        JSONObject lightsResponse = Network.performGetRequest(buildApiUrl("lights"));
        List<ILightID> result = new ArrayList<>();

        Iterator<String> lightIdIterator = lightsResponse.keys();
        while (lightIdIterator.hasNext()) {
            final String lightId = lightIdIterator.next();

            JSONObject lightInfo = (JSONObject) lightsResponse.get(lightId);
            String modelId = (String) lightInfo.get("modelid");
            String uniqueId = (String) lightInfo.get("uniqueid");
            uniqueId = uniqueId.substring(0, 23);

            if (modelId.contains("LLC")) {
                result.add(new LightIDImpl(Integer.parseInt(lightId), uniqueId));
            }
        }

        return result;
    }


    private void includeLightsIntoGroup_internal(List<ILightID> lights, Integer targetGroupId) throws JSONException {
        JSONArray lightsArray = new JSONArray();
        for (ILightID id : lights) {
            lightsArray.put(id.getLightId().toString());
        }

        JSONObject request = new JSONObject();
        request.put("lights", lightsArray);

        JSONArray response = Network.performUpdateRequest(buildApiUrl("groups/" + targetGroupId), request, Network.UPDATE_REQUEST.PUT);
        getSuccessNode(response);
    }


    private void setLightsCoordinates_internal(Integer groupId, List<ILightCoordinate> lightCoordinates) throws JSONException {
        JSONObject locationsNode = new JSONObject();

        for (ILightCoordinate coordinate: lightCoordinates) {
            JSONArray coordArray = new JSONArray();
            coordArray.put(coordinate.getX());
            coordArray.put(coordinate.getY());
            coordArray.put(0);

            locationsNode.put(String.valueOf(coordinate.getLightId()), (Object) coordArray);
        }
        JSONObject request = new JSONObject();
        request.put("locations", (Object) locationsNode);

        JSONArray response = Network.performUpdateRequest(buildApiUrl("groups/" + groupId), request, Network.UPDATE_REQUEST.PUT);
        Assert.assertNotNull("Update locations response is null", response);

        for (int i = 0; i < response.length(); ++i) {
            JSONObject lightNode = response.getJSONObject(i);
            Assert.assertNotNull("Invalid response format", lightNode);

            JSONObject successNode = (JSONObject) lightNode.get("success");
            Assert.assertNotNull("Creating group was not successful", successNode);
        }
    }


    private void cleanUpUser_internal() throws JSONException {
        JSONArray response = Network.performDeleteRequest(buildApiUrl("config/whitelist/" + _userName));
        getSuccessNode(response);
    }

    Integer createEntertainmentGroup() throws JSONException {
        JSONObject request = new JSONObject();
        request.put("type", "Entertainment");
        JSONArray lightsArray = new JSONArray();
        request.put("lights", (Object) lightsArray);
        request.put("class", "TV");

        JSONArray response = Network.performUpdateRequest(buildApiUrl("groups"), request, Network.UPDATE_REQUEST.POST);
        JSONObject successNode = (JSONObject) getSuccessNode(response);

        String groupIDValue = (String)successNode.get("id");
        Assert.assertNotNull("Can not find group id", groupIDValue);

        return Integer.parseInt(groupIDValue);
    }

    List<Integer> getAllEntertainmentGroups() throws JSONException {
        JSONObject groupsResponse = Network.performGetRequest(buildApiUrl("groups"));
        Assert.assertNotNull("Groups JSON object is NULL", groupsResponse);
        List<Integer> result = new ArrayList<>();

        Iterator<String> groupIdIterator = groupsResponse.keys();
        while (groupIdIterator.hasNext()) {
            final String groupId = groupIdIterator.next();
            JSONObject group = (JSONObject) groupsResponse.get(groupId);

            String groupType = (String) group.get("type");
            if (groupType.equals("Entertainment")) {
                result.add(Integer.parseInt(groupId));
            }
        }

        return result;
    }

    String buildApiUrl(final String urlTail) {
        StringBuilder urlBuilder = new StringBuilder();
        urlBuilder.append("http://")
                .append(_ip4Address)
                .append(":")
                .append(_tcpPort)
                .append("/api/")
                .append(_userName)
                .append("/")
                .append(urlTail);

        return urlBuilder.toString();
    }

    Object getSuccessNode(final JSONArray response) throws JSONException {
        Assert.assertNotNull("Create/Update/Delete response is null", response);

        JSONObject responseRoot = (JSONObject) response.get(0);
        Assert.assertNotNull("Wrong structure of incoming response", responseRoot);

        Object successNode = responseRoot.get("success");
        Assert.assertNotNull("No \"success\" node is found", successNode);

        return successNode;
    }

    private final String _userName;
    private final String _clientKey;
    private final String _ip4Address;
    private final String _bridgeId;
    private final String _tcpPort;
}
